hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java

```diff
@@ -338,49 +338,53 @@ private ResponseCodeEnum setIsFrozen(
 
 	@Override
 	public ResponseCodeEnum adjustBalance(AccountID aId, TokenID tId, long adjustment) {
-		return sanityChecked(aId, tId, token -> tryAdjustment(aId, tId, adjustment));
+		return sanityChecked(aId, null, tId, token -> tryAdjustment(aId, tId, adjustment));
 	}
 
 	@Override
-	public ResponseCodeEnum adjustBalance(AccountID senderAId, AccountID receiverAId, TokenID tId, long serialNumber) {
-		var validity = sanityChecked(senderAId, receiverAId, tId, token -> tryAdjustment(senderAId, receiverAId, tId, serialNumber));
-		if (validity == OK) {
-			var nftId = NftID.newBuilder().setTokenID(tId).setSerialNumber(serialNumber).build();
-			if (nftExists(nftId)) {
-				get(nftId).setOwner(EntityId.fromGrpcAccountId(receiverAId));
-
-				adjustOwnedNfts(senderAId, false);
-				adjustOwnedNfts(receiverAId, true);
-			} else {
-				validity = INVALID_NFT_ID;
+	public ResponseCodeEnum changeOwner(NftId nftId, AccountID from, AccountID to) {
+		final var tId = nftId.tokenId();
+		return sanityChecked(from, to, tId, token -> {
+			if (!nftsLedger.exists(nftId)) {
+				return INVALID_NFT_ID;
 			}
-		}
 
-		return validity;
-	}
+			final var fromFreezeAndKycValidity = checkRelFrozenAndKycProps(from, tId);
+			if (fromFreezeAndKycValidity != OK) {
+				return fromFreezeAndKycValidity;
+			}
+			final var toFreezeAndKycValidity = checkRelFrozenAndKycProps(to, tId);
+			if (toFreezeAndKycValidity != OK) {
+				return toFreezeAndKycValidity;
+			}
 
-	public boolean nftExists(final NftID id) {
-		return uniqueTokenSupplier.get().containsKey(fromNftID(id));
-	}
+			final var owner = (EntityId) nftsLedger.get(nftId, OWNER);
+			if (!owner.matches(from)) {
+				return SENDER_DOES_NOT_OWN_NFT_SERIAL_NO;
```

Comment:
I am not sure how we can get to this use-case. In `HederaSigningOrder` -> `cryptoTransfer` we are requiring the `sender` account to provide signature for the transfer.

Actually, if another TX transfers the NFT between the signature expansion and the `handleTransaction` we might end up in this situation. Is this the reason why you've added this check?

Reply:
This check isn't done anywhere in `expandSignatures` afaict, so any `CryptoTransfer` that tries to move an NFT from a non-owner account should get here, right?

Reply:
We updated the signing order to add the sender account for activation here:
https://github.com/hashgraph/hedera-services/blob/9bb3eeaa0f5befa69bc78f6bc2b9367f0651feca/hedera-node/src/main/java/com/hedera/services/sigs/order/HederaSigningOrder.java#L561